View3D/Rendering/RenderItems/ConeRenderItem.cs

```diff
@@ -0,0 +1,126 @@
+            vectorP.Normalize();
+            
+            Vector3 planeVectorP = Vector3.Cross(normal, Vector3.Cross(vectorP, normal));
+            Vector3 planeVectorPN = Vector3.Cross(vectorP, normal);
+            planeVectorP.Normalize();
+            planeVectorPN.Normalize();
+
+            var circleSteps = new List<float>();
+            var angleStep = MathF.PI / 60;
+            for (float angle = angleStep; angle < MathF.PI; angle += angleStep)
+            {
+                if (angle < halfAngle)
+                {
+                    circleSteps.Add(angle);
+                }
+            }
+            int circleCount = circleSteps.Count;
+            
+            var fullCircle = 2 * MathF.PI;
+            var steps = 20;
+            var stepsSize = fullCircle / steps;
+
+            if (!(Math.Abs(_coneAngle - 360.0f) < 0.001f))
+            {
+                _lastCircle = new VertexPositionColor[steps + 1];
+                _rays = new VertexPositionColor[2 * (steps + 1)];
+                for (int j = 0; j < steps + 1; j++)
+                {
+                    float angle = stepsSize * j;
+                    Vector3 rotatedVector = normal * MathF.Cos(halfAngle) + planeVectorP * MathF.Sin(halfAngle) * MathF.Sin(angle) + planeVectorPN * MathF.Cos(angle) * MathF.Sin(halfAngle);
+                    rotatedVector.Normalize();
+                    _lastCircle[j] = new VertexPositionColor(_startPos + rotatedVector * distance, _colour);
+                    _rays[2*j] = new VertexPositionColor(_startPos, _colour);
+                    _rays[2*j+1] = new VertexPositionColor(_lastCircle[j].Position, _colour);
+                }
+            }
+            
+            // Array of {#circleCount} circles of {#steps+1} points each
+            _circles = new VertexPositionColor[circleCount][];
+            for (int i = 0; i < circleCount; i++)
+            {
+                _circles[i] = new VertexPositionColor[steps + 1];
+                float sectorAngle = circleSteps[i];
+                
+                for (int j = 0; j < steps + 1; j++)
+                {
+                    float angle = stepsSize * j;
+                    Vector3 rotatedVector = normal * MathF.Cos(sectorAngle) + planeVectorP * MathF.Sin(sectorAngle) * MathF.Sin(angle) + planeVectorPN * MathF.Cos(angle) * MathF.Sin(sectorAngle);
+                    rotatedVector.Normalize();
+                    _circles[i][j] = new VertexPositionColor(_startPos + rotatedVector * distance, _colour);
+                }
+            }
+        }
+
+        public void Draw(GraphicsDevice device, CommonShaderParameters parameters)
+        {
+            _shader.Parameters["View"].SetValue(parameters.View);
+            _shader.Parameters["Projection"].SetValue(parameters.Projection);
+            _shader.Parameters["World"].SetValue(ModelMatrix);
+        
+            foreach (var pass in _shader.CurrentTechnique.Passes)
+            {
+                pass.Apply();
```

Comment:
I think you can change this to just upload a list of points to the card and save the loop. 
Have a look at LineMeshRender.cs

if applicable, could this be solved by adding a new function like the void AddCube(Matrix transform, Color color) function?

Reply:
Whoops, if-statement should be outside of the inner for-loop, just noticed.

> I think you can change this to just upload a list of points to the card and save the loop.
I can replace `_circles `with 1d array with the size twice as large and pass it to `device.DrawUserPrimitives` with `PrimitiveType.LineList `argument. Will it be faster?

Reply:
device.DrawXXX is super slow. Not a huge issue for a simple case like this, but I think its among the slowest calls in a game engine and one typically goes to (extreme) lengths to minimize the call count.